File: Habitica/src/main/java/com/habitrpg/android/habitica/models/TutorialStep.java
package com.habitrpg.android.habitica.models;

import com.habitrpg.android.habitica.HabitDatabase;
import com.habitrpg.android.habitica.models.user.Flags;
import com.raizlabs.android.dbflow.annotation.Column;
import com.raizlabs.android.dbflow.annotation.PrimaryKey;
import com.raizlabs.android.dbflow.annotation.Table;
import com.raizlabs.android.dbflow.structure.BaseModel;

import java.util.Date;

import io.realm.RealmObject;

public class TutorialStep extends RealmObject {

    public Flags flags;
    @PrimaryKey
    private String key;
    private String tutorialGroup, identifier;
    private boolean wasCompleted;
    private Date displayedOn;

    public String getKey() {
        return key;
    }

    public void setKey(String key) {
        this.key = key;
    }

    public String getTutorialGroup() {
        return tutorialGroup;
    }

    public void setTutorialGroup(String group) {
        this.tutorialGroup = group;
        this.key = group + "_" + this.identifier;
    }

    public String getIdentifier() {
        return identifier;
    }

    public void setIdentifier(String identifier) {
        this.identifier = identifier;
        this.key = this.tutorialGroup + "_" + identifier;
    }

    public boolean getWasCompleted() {
        return wasCompleted;
    }

    public void setWasCompleted(boolean wasCompleted) {
        this.wasCompleted = wasCompleted;
    }

    public Date getDisplayedOn() {
        return displayedOn;
    }

    public void setDisplayedOn(Date displayedOn) {
        this.displayedOn = displayedOn;
    }

    public boolean shouldDisplay() {
        return  !this.getWasCompleted() && (this.getDisplayedOn() == null || (new Date().getTime() - this.getDisplayedOn().getTime()) > 86400000);
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/models/TutorialStep.kt
package com.habitrpg.android.habitica.models

import io.realm.RealmModel
import io.realm.RealmObject
import io.realm.annotations.RealmClass
import java.util.Date

@RealmClass(embedded = true)
open class TutorialStep : RealmObject(), BaseMainObject {

    var key: String? = null
    var tutorialGroup: String? = null
        set(group) {
            field = group
            this.key = group + "_" + this.identifier
        }
    var identifier: String? = null
        set(identifier) {
            field = identifier
            this.key = this.tutorialGroup + "_" + identifier
        }
    var wasCompleted: Boolean = false
    var displayedOn: Date? = null

    val shouldDisplay: Boolean
        get() {
            return !this.wasCompleted && (
                this.displayedOn == null || Date().time - (
                    displayedOn?.time
                        ?: 0
                    ) > 86400000
                )
        }
    override val realmClass: Class<out RealmModel>
        get() = TutorialStep::class.java
    override val primaryIdentifier: String?
        get() = key
    override val primaryIdentifierName: String
        get() = "key"

    val flagPath: String
        get() = "flags.tutorial.$tutorialGroup.$identifier"

    val linkFAQ : Boolean
        get() {
            return identifier == "party"
        }
}
